Mathlib/Topology/Algebra/LinearTopology.lean

```diff
@@ -261,6 +271,96 @@ theorem _root_.IsCentralScalar.isLinearTopology_iff [Module Rᵐᵒᵖ M] [IsCen
+
+lemma _root_.Submodule.isOpen_leAddSubgroup_of_compactSpace
+    {G : AddSubgroup M} (hG : IsOpen (X := M) G) :
+    IsOpen (X := M) (Submodule.leAddSubgroup R G) := by
+  refine isOpen_iff_mem_nhds.mpr fun x hx ↦ ?_
+  suffices ∀ s : Set R, IsCompact s → ∃ V ∈ 𝓝 x, ∀ a ∈ s, ∀ b ∈ V, a • b ∈ G by
+    obtain ⟨V, hV, H⟩ := this Set.univ isCompact_univ
+    refine (𝓝 x).mem_of_superset hV ?_
+    simpa [Set.subset_def, Submodule.mem_leAddSubgroup, forall_comm (α := R)] using H
+  intros s hs
+  refine hs.induction_on ?_ ?_ ?_ ?_
+  · simp only [Set.mem_empty_iff_false, IsEmpty.forall_iff, implies_true, and_true]
+    exact ⟨Set.univ, Filter.univ_mem⟩
+  · rintro s₁ s₂ hs₁s₂ ⟨V, hV, H⟩
+    exact ⟨V, hV, fun a ha b hb ↦ H a (hs₁s₂ ha) b hb⟩
+  · rintro s₁ s₂ ⟨V₁, hV₁, H₁⟩ ⟨V₂, hV₂, H₂⟩
+    exact ⟨_, Filter.inter_mem hV₁ hV₂, fun a ha b hb ↦ ha.elim (H₁ a · b hb.1) (H₂ a · b hb.2)⟩
+  · intro a ha
+    have := continuous_smul.tendsto (a, x) (hG.mem_nhds (hx _))
+    simp only [nhds_prod_eq, Filter.mem_map, Filter.mem_prod_iff] at this
+    obtain ⟨t₁, ht₁, t₂, ht₂, H⟩ := this
+    exact ⟨t₁, mem_nhdsWithin_of_mem_nhds ht₁, t₂, ht₂, fun a ha b hb ↦ @H (a, b) (by simp [*])⟩
+
+instance (priority := low) [NonarchimedeanAddGroup M] :
+    IsLinearTopology R M := by
+  refine isLinearTopology_iff_exists_submodule_isOpen_subset.mpr fun s hs h0s ↦ ?_
+  obtain ⟨V, hV⟩ := NonarchimedeanAddGroup.is_nonarchimedean s (hs.mem_nhds h0s)
+  exact ⟨.leAddSubgroup R V.1, Submodule.isOpen_leAddSubgroup_of_compactSpace V.2,
+    subset_trans (Submodule.leAddSubgroup_le _ _) hV⟩
+
+lemma _root_.isLinearTopology_iff_nonarchimedeanAddGroup_of_compactSpace [IsTopologicalAddGroup M] :
+    IsLinearTopology R M ↔ NonarchimedeanAddGroup M := by
+  refine ⟨fun _ ↦ IsLinearTopology.nonarchimedeanAddGroup R M, fun _ ↦ inferInstance⟩
+
+end CompactSpace
+
+section restrictScalars
+
+variable (R S M)
+variable [Ring S] [Module S M]
+
+lemma restrictScalars [SMul R S] [IsScalarTower R S M] [hS : IsLinearTopology S M] :
+    IsLinearTopology R M := by
+  rw [isLinearTopology_iff_hasBasis_submodule] at hS ⊢
+  exact hS.to_hasBasis' (fun N hN ↦ ⟨N.restrictScalars R, hN, by simp⟩) (by simp)
+
+lemma of_restrictScalars [Module R S] [IsScalarTower R S M] [Module.Finite R S]
+    [IsLinearTopology R M] [ContinuousAdd M] [ContinuousConstSMul S M] :
+    IsLinearTopology S M := by
+  rw [isLinearTopology_iff_exists_submodule_isOpen_subset]
+  intro s hs h0s
+  obtain ⟨N, hN, hNs⟩ := (IsLinearTopology.hasBasis_open_submodule R).mem_iff.mp (hs.mem_nhds h0s)
+  refine ⟨_, ?_, subset_trans (Submodule.leAddSubgroup_le _ N.toAddSubgroup) hNs⟩
+  obtain ⟨s, hs⟩ := Module.Finite.fg_top (R := R) (M := S)
+  convert isOpen_biInter_finset (s := s) (fun i _ ↦ (continuous_const_smul i).isOpen_preimage _ hN)
+  ext
+  simp [Submodule.mem_leAddSubgroup_iff_of_span_eq_top _ hs]
+
+lemma iff_restrictScalars [Module R S] [IsScalarTower R S M] [Module.Finite R S]
```

Comment:
There could be an additional `iff` lemma that builds on the above result for compact modules.